fem/geom.hpp

```diff
@@ -97,9 +97,9 @@ class MFEM_EXPORT Geometry
 
    const DenseMatrix &GetGeomToPerfGeomJac(int GeomType) const
    { return *GeomToPerfGeomJac[GeomType]; }
-   DenseMatrix *GetPerfGeomToGeomJac(int GeomType)
-   { return PerfGeomToGeomJac[GeomType]; }
-   void GetPerfPointMat(int GeomType, DenseMatrix &pm);
+   const DenseMatrix &GetPerfGeomToGeomJac(int GeomType) const
```

Comment:
Is this an acceptable modification? It looks like the initial code was actually a mistake which was inconsitent with the above `GetGeomToPerfGeomJac` function and lacked `const`-correctness. It's not a "sneaky change" in the sense that anyone building their code using this function when they update to a new version of MFEM will get a compiler error and can make the straightforward change in their application.

Reply:
I think for now we mostly gathered inconsistencies (and hence possible breaking changes) here https://github.com/mfem/mfem/issues/1557 . Technically it is breaking in the sense of SemVer, but I would agree that we enter a grey zone here. We already had silent breaking changes in the past MFEM 4.x releases (e.g. by being stricter about const correctness of public methods), but I assume that the resulting issues were, same as here and as you suggest, easy to fix for downstream users and library devs.

Reply:
Agreed, I'll leave as is for now unless there is a significant reason not to.

Reply:
As discussed I've reverted the breaking change by adding the previous methods with an `MFEM_DEPRECATED` attribute in [5f85120](https://github.com/mfem/mfem/pull/3846/commits/5f85120da441d69eaaac10d08747be4b99c42c47).

Reply:
Looking at the code, the method `GetPerfGeomToGeomJac` was meant to return a pointer since in the case when this Jacobian is the identity, it will return `NULL` (for performance considerations). If we dereference the `NULL` pointers, we create undefined behavior. For the other method, `GetGeomToPerfGeomJac`, I see that all geometries of interest initialize the internal array `GeomToPerfGeomJac` with non-`NULL` pointers, so there should be no issues with undefined behavior when dereferencing.

With this in mind, I think it is probably best to keep the code as in master, unless this somehow affects thread-safety which is the main goal of the PR. What do you think?

Reply:
That makes sense! Thanks @v-dobrev. Fixed in https://github.com/mfem/mfem/pull/3846/commits/c9e112c19201810e8dcdecf9418d7e252d65098f. I've at least changed the message to return a pointer to `const` so the method is `const` which is appropriate to be used simultaneously by threads.